File: java/linkedlist/ReverseList.java
/**
 * Q6. Reverse Linked List
 * 
 * Problem:
 * Given the head of a singly linked list, reverse the list, and return the reversed list.
 * 
 * Example:
 * Input: head = [1,2,3,4,5]
 * Output: [5,4,3,2,1]
 * 
 * Approach:
 * 1. Use three pointers: prev, current, and next
 * 2. Initialize prev as null and current as head
 * 3. For each node:
 *    - Store next node
 *    - Reverse current node's pointer
 *    - Move prev and current one step forward
 * 4. Return prev as new head
 * 
 * Time Complexity: O(N) where N is the length of the linked list
 * Space Complexity: O(1) as we only use three pointers
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class ReverseList {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;
        
        while (current != null) {
            // Store next node
            ListNode nextNode = current.next;
            
            // Reverse current node's pointer
            current.next = prev;
            
            // Move prev and current one step forward
            prev = current;
            current = nextNode;
        }
        
        // prev will be the new head
        return prev;
    }

    // Recursive approach
    public ListNode reverseListRecursive(ListNode head) {
        // Base case: empty list or single node
        if (head == null || head.next == null) {
            return head;
        }
        
        // Recursive call to reverse the rest of the list
        ListNode reversedList = reverseListRecursive(head.next);
        
        // Reverse the current node
        head.next.next = head;
        head.next = null;
        
        return reversedList;
    }

    // Main method for testing
    public static void main(String[] args) {
        // Create test linked list: 1->2->3->4->5
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        // Reverse the list iteratively
        ReverseList solution = new ReverseList();
        ListNode reversed = solution.reverseList(head);

        // Print the reversed list
        System.out.print("Reversed List (Iterative): ");
        while (reversed != null) {
            System.out.print(reversed.val + " -> ");
            reversed = reversed.next;
        }
        System.out.println("null");

        // Create another list for recursive approach
        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(2);
        head2.next.next = new ListNode(3);
        head2.next.next.next = new ListNode(4);
        head2.next.next.next.next = new ListNode(5);

        // Reverse the list recursively
        ListNode reversedRecursive = solution.reverseListRecursive(head2);

        // Print the recursively reversed list
        System.out.print("Reversed List (Recursive): ");
        while (reversedRecursive != null) {
            System.out.print(reversedRecursive.val + " -> ");
            reversedRecursive = reversedRecursive.next;
        }
        System.out.println("null");
    }
} 

File: javascript/linkedlist/ReverseList.js
/**
 * Q9. Reverse Linked List
 * 
 * Problem:
 * Given the head of a singly linked list, reverse the list, and return the reversed list.
 * 
 * Example:
 * Input: head = [1,2,3,4,5]
 * Output: [5,4,3,2,1]
 * 
 * Approach:
 * Iterative approach:
 * 1. Use three pointers: prev, current, and next
 * 2. Initialize prev as null and current as head
 * 3. For each node:
 *    - Store next node
 *    - Reverse current node's pointer
 *    - Move prev and current one step forward
 * 4. Return prev as new head
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class ReverseList {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        let prev = null;
        let current = head;

        while (current) {
            const next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }

    /**
     * Recursive approach to reverse linked list
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseListRecursive(head) {
        if (!head || !head.next) return head;

        const newHead = this.reverseListRecursive(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }

    // Helper method to create a linked list
    static createList(values) {
        if (!values || values.length === 0) return null;
        
        const head = new ListNode(values[0]);
        let current = head;
        
        for (let i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;
        }
        
        return head;
    }

    // Helper method to print the list
    static printList(head) {
        const values = [];
        let current = head;
        
        while (current) {
            values.push(current.val);
            current = current.next;
        }
        
        console.log(values.join(' -> '));
    }
}

// Test cases
function runTests() {
    const solution = new ReverseList();

    // Test case 1: Normal list
    console.log("Test case 1: Normal list");
    const list1 = ReverseList.createList([1, 2, 3, 4, 5]);
    console.log("Original list:");
    ReverseList.printList(list1);
    console.log("Reversed list (iterative):");
    ReverseList.printList(solution.reverseList(list1));
    console.log("Reversed list (recursive):");
    ReverseList.printList(solution.reverseListRecursive(ReverseList.createList([1, 2, 3, 4, 5])));

    // Test case 2: Two node list
    console.log("\nTest case 2: Two node list");
    const list2 = ReverseList.createList([1, 2]);
    console.log("Original list:");
    ReverseList.printList(list2);
    console.log("Reversed list (iterative):");
    ReverseList.printList(solution.reverseList(list2));
    console.log("Reversed list (recursive):");
    ReverseList.printList(solution.reverseListRecursive(ReverseList.createList([1, 2])));

    // Test case 3: Single node list
    console.log("\nTest case 3: Single node list");
    const list3 = ReverseList.createList([1]);
    console.log("Original list:");
    ReverseList.printList(list3);
    console.log("Reversed list (iterative):");
    ReverseList.printList(solution.reverseList(list3));
    console.log("Reversed list (recursive):");
    ReverseList.printList(solution.reverseListRecursive(ReverseList.createList([1])));

    // Test case 4: Empty list
    console.log("\nTest case 4: Empty list");
    const list4 = null;
    console.log("Original list: null");
    console.log("Reversed list (iterative):");
    ReverseList.printList(solution.reverseList(list4));
    console.log("Reversed list (recursive):");
    ReverseList.printList(solution.reverseListRecursive(null));

    // Test case 5: List with negative numbers
    console.log("\nTest case 5: List with negative numbers");
    const list5 = ReverseList.createList([-5, -4, -3, -2, -1]);
    console.log("Original list:");
    ReverseList.printList(list5);
    console.log("Reversed list (iterative):");
    ReverseList.printList(solution.reverseList(list5));
    console.log("Reversed list (recursive):");
    ReverseList.printList(solution.reverseListRecursive(ReverseList.createList([-5, -4, -3, -2, -1])));
}

// Run the tests
runTests(); 